torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -63,197 +64,154 @@ static int64_t count_specified_dimensions(PyObject* index) {
-  if (jit::tracer::isTracing() && THPVariable_Check(index)) {
-    auto& var = THPVariable_Unpack(index);
-    jit::tracer::ArgumentStash::stashValue(std::string("index"), 1, var, jit::IntType::get());
-  }
-
-  int64_t unpacked_index = THPUtils_unpackLong(index);
-  if (unpacked_index == 0 && dim == 0 && self.dim() == 0) {
-    throw IndexError(
-        "invalid index of a 0-dim tensor. "
-        "Use tensor.item() to convert a 0-dim tensor to a Python number");
-  }
-  int64_t size = self.size(dim);
-  if (unpacked_index < -size || unpacked_index >= size) {
-    throw IndexError("index %lld is out of bounds for dimension %lld with size %lld",
-      unpacked_index, real_dim, size);
-  }
-  // if the index is negative, do not normalize it because that would fix the index
-  // on the current tensor size in the tracer.
-  // aten::select also works on negative indices
-  return self.select(dim, unpacked_index);
-}
-
-static Variable sequenceToVariable(c10::DispatchKey dispatch_key, PyObject* seq) {
+static inline Variable sequenceToVariable(c10::DispatchKey dispatch_key, PyObject* seq) {
   return torch::utils::indexing_tensor_from_data(dispatch_key, kLong, c10::nullopt, seq);
 }
 
-static Variable valueToTensor(c10::TensorOptions options, PyObject* value) {
+static inline Variable valueToTensor(c10::TensorOptions options, PyObject* value, const at::Device& device) {
   if (THPVariable_Check(value)) {
     return reinterpret_cast<THPVariable*>(value)->cdata;
   }
   at::AutoNonVariableTypeMode guard;
   if (THPUtils_checkLong(value) || PyBool_Check(value)) {
-    return at::scalar_tensor(Scalar(THPUtils_unpackLong(value)), options);
+    return at::indexing::scalarToTensor(Scalar(THPUtils_unpackLong(value)), options, device);
   }
   if (PyFloat_Check(value)) {
-    return at::scalar_tensor(Scalar(THPUtils_unpackDouble(value)), options);
+    return at::indexing::scalarToTensor(Scalar(THPUtils_unpackDouble(value)), options, device);
   }
   throw TypeError(
     "can't assign a %s to a %s",
     Py_TYPE(value)->tp_name,
     torch::utils::options_to_string(options).c_str());
 }
 
-static Variable boolToIndexingTensor(const Variable& self, bool value) {
-  // booleans add a dimension of size 1. true indexes this dimension as if 0:, false as empty.
-  if (value) {
-    return at::zeros({1}, self.options().dtype(kLong));
-  } else {
-    return at::empty({0}, self.options().dtype(kLong));
+static inline void extractTensorsFromSlice(
+  PyObject* obj,
+  Tensor* start_tensor_ptr,
+  Tensor* stop_tensor_ptr,
+  Tensor* step_tensor_ptr) {
+  PySliceObject* sliceobj = (PySliceObject*)obj;
+  if (THPVariable_Check(sliceobj->start)) {
+    (*start_tensor_ptr) = THPVariable_Unpack(sliceobj->start);
+  }
+  if (THPVariable_Check(sliceobj->stop)) {
+    (*stop_tensor_ptr) = THPVariable_Unpack(sliceobj->stop);
+  }
+  if (THPVariable_Check(sliceobj->step)) {
+    (*step_tensor_ptr) = THPVariable_Unpack(sliceobj->step);
+  }
```

Comment:
This function is always called in exactly the same way with the exact same functions around it:

```
          Tensor start_tensor, stop_tensor, step_tensor;
          extractTensorsFromSlice(obj, &start_tensor, &stop_tensor, &step_tensor);
          recordSliceTrace(start_tensor, stop_tensor, step_tensor);
```

Just combine this all into a single function...

Reply:
In fact, the lines above also seem replicated everywhere:

```

    Py_ssize_t start, stop, step;
    if (!THPUtils_unpackSlice(index, &start, &stop, &step)) {
      throw python_error();
    }
```